File: FCL/src/main/java/com/tungsten/fcl/ui/manage/WorldListAdapter.java
package com.tungsten.fcl.ui.manage;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.tungsten.fcl.R;
import com.tungsten.fclcore.fakefx.beans.InvalidationListener;
import com.tungsten.fclcore.fakefx.beans.property.ListProperty;
import com.tungsten.fclcore.fakefx.beans.property.SimpleListProperty;
import com.tungsten.fclcore.fakefx.collections.FXCollections;
import com.tungsten.fcllibrary.component.FCLAdapter;
import com.tungsten.fcllibrary.component.view.FCLImageButton;
import com.tungsten.fcllibrary.component.view.FCLLinearLayout;
import com.tungsten.fcllibrary.component.view.FCLTextView;

public class WorldListAdapter extends FCLAdapter {

    private final ListProperty<WorldListItem> listProperty = new SimpleListProperty<>(FXCollections.observableArrayList());

    public ListProperty<WorldListItem> listProperty() {
        return listProperty;
    }

    public WorldListAdapter(Context context) {
        super(context);

        listProperty.addListener((InvalidationListener) observable -> {
            notifyDataSetChanged();
        });
    }

    private static class ViewHolder {
        FCLLinearLayout parent;
        FCLTextView name;
        FCLTextView description;
        FCLImageButton datapack;
        FCLImageButton export;
        FCLImageButton delete;
    }

    @Override
    public int getCount() {
        return listProperty.getSize();
    }

    @Override
    public Object getItem(int i) {
        return listProperty.get(i);
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        final ViewHolder viewHolder;
        if (view == null) {
            viewHolder = new ViewHolder();
            view = LayoutInflater.from(getContext()).inflate(R.layout.item_world, null);
            viewHolder.parent = view.findViewById(R.id.parent);
            viewHolder.name = view.findViewById(R.id.name);
            viewHolder.description = view.findViewById(R.id.description);
            viewHolder.datapack = view.findViewById(R.id.datapack);
            viewHolder.export = view.findViewById(R.id.export);
            viewHolder.delete = view.findViewById(R.id.delete);
            view.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) view.getTag();
        }
        WorldListItem worldListItem = listProperty.get(i);
        viewHolder.parent.setOnClickListener(v -> worldListItem.showInfo());
        viewHolder.name.stringProperty().bind(worldListItem.titleProperty());
        viewHolder.description.stringProperty().bind(worldListItem.subtitleProperty());
        viewHolder.datapack.setOnClickListener(v -> worldListItem.manageDatapacks());
        viewHolder.export.setOnClickListener(v -> worldListItem.export());
        viewHolder.delete.setOnClickListener(v -> worldListItem.delete());
        return view;
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/manage/WorldListAdapter.kt
package com.tungsten.fcl.ui.manage

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mio.ui.adapter.ViewHolder
import com.tungsten.fcl.databinding.ItemWorldBinding
import com.tungsten.fcl.util.FlowList
import com.tungsten.fclcore.util.flow.FlowSubscriptions

class WorldListAdapter(private val context: Context) :
    RecyclerView.Adapter<ViewHolder>() {

    private val listProperty: FlowList<WorldListItem> = FlowList()

    fun listProperty(): FlowList<WorldListItem> {
        return listProperty
    }

    init {
        FlowSubscriptions.subscribe(listProperty.flow()) { _: List<WorldListItem> ->
            notifyDataSetChanged()
        }
    }

    // 视图回收重绑时持有的订阅（存于 itemView.tag），重绑前 cancel 旧订阅
    //（对齐原 bind 重复调用先解绑的语义）。
    private class BindingSubscriptions {
        var title: FlowSubscriptions.Subscription? = null
        var subtitle: FlowSubscriptions.Subscription? = null
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        return ViewHolder(
            ItemWorldBinding.inflate(
                LayoutInflater.from(context),
                parent,
                false
            ).root
        )
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val binding = ItemWorldBinding.bind(holder.itemView)
        val worldListItem = listProperty.get()[position]
        val subscriptions =
            holder.itemView.tag as? BindingSubscriptions ?: BindingSubscriptions().also {
                holder.itemView.tag = it
            }
        binding.parent.setOnClickListener { worldListItem.showInfo() }
        subscriptions.title?.cancel()
        binding.name.stringFlow().value = worldListItem.titleFlow().value
        subscriptions.title = FlowSubscriptions.subscribe(worldListItem.titleFlow()) { v ->
            binding.name.stringFlow().value = v
        }
        subscriptions.subtitle?.cancel()
        binding.description.stringFlow().value = worldListItem.subtitleFlow().value
        subscriptions.subtitle = FlowSubscriptions.subscribe(worldListItem.subtitleFlow()) { v ->
            binding.description.stringFlow().value = v
        }
        binding.datapack.setOnClickListener { worldListItem.manageDatapacks() }
        binding.export.setOnClickListener { worldListItem.export() }
        binding.delete.setOnClickListener { worldListItem.delete() }
    }

    override fun getItemCount(): Int {
        return listProperty.size()
    }
}
